### PR TITLE
HPCC-31457 Use PCRE2 for native UTF-8 regex

### PR DESCRIPTION
**Pull request by dcamper**
## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [x] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [x] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [x] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [ ] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [x] I have added tests to cover my changes.
  - [x] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [x] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

**Review by jackdelv — APPROVED** [review]
Looks good.

**Comment by AttilaVamos** [comment]
There are some regression failures (on all engines):
```
599. Test: regex2.ecl
    eclcc: W20240417-140558-4.cpp(399,64): Error C6003: cannot convert â€˜UChar*â€™ {aka â€˜short unsigned int*â€™} to â€˜char*&â€™
    eclcc: W20240417-140558-4.cpp(405,64): Error C6003: cannot convert â€˜UChar*â€™ {aka â€˜short unsigned int*â€™} to â€˜char*&â€™
    eclcc: W20240417-140558-4.cpp(411,64): Error C6003: cannot convert â€˜UChar*â€™ {aka â€˜short unsigned int*â€™} to â€˜char*&â€™
    eclcc: W20240417-140558-4.cpp(417,64): Error C6003: cannot convert â€˜UChar*â€™ {aka â€˜short unsigned int*â€™} to â€˜char*&â€™
    eclcc: W20240417-140558-4.cpp(423,64): Error C6003: cannot convert â€˜UChar*â€™ {aka â€˜short unsigned int*â€™} to â€˜char*&â€™
    eclcc: W20240417-140558-4.cpp(440,64): Error C6003: cannot convert â€˜UChar*â€™ {aka â€˜short unsigned int*â€™} to â€˜char*&â€™
    eclcc: W20240417-140558-4.cpp(473,64): Error C6003: cannot convert â€˜UChar*â€™ {aka â€˜short unsigned int*â€™} to â€˜char*&â€™
    eclccserver: (0,0): Error C3000: Compile/Link failed for W20240417-140558-4 (see W20240417-140558-4.cc.log for details)
    8 error(s), 7 warning(s)
    
    600. Test: regex3.ecl
    eclcc: W20240417-140559.cpp(188,62): Error C6003: cannot convert â€˜UChar*â€™ {aka â€˜short unsigned int*â€™} to â€˜char*&â€™
    eclcc: W20240417-140559_1.cpp(78,49): Error C6003: cannot convert â€˜UChar*â€™ {aka â€˜short unsigned int*â€™} to â€˜char*&â€™
    eclcc: W20240417-140559_1.cpp(90,49): Error C6003: cannot convert â€˜UChar*â€™ {aka â€˜short unsigned int*â€™} to â€˜char*&â€™
    eclccserver: (0,0): Error C3000: Compile/Link failed for W20240417-140559 (see W20240417-140559.cc.log for details)
    4 error(s), 3 warning(s)
    
    603. Test: regexfindset.ecl
    eclccserver: (0,0): Error C9999: eclcc killed - likely to be out of memory - see compile log for details
    1 error(s), 0 warning(s)
```
```
W20240417-142609-5.cpp: In member function ‘virtual int MyEclProcess::perform(IGlobalCodeContext*, unsigned int)’:
W20240417-142609-5.cpp:188:62: error: cannot convert ‘UChar*’ {aka ‘short unsigned int*’} to ‘char*&’
  188 |                                 regexI->replace(vH,vG.refustr(),22U,((char *)"the cat sat on the mat"),3U,((char *)"$1p"));
      |                                                    ~~~~~~~~~~^~
      |                                                              |
      |                                                              UChar* {aka short unsigned int*}
In file included from /opt/HPCCSystems/componentfiles/cl/include/eclinclude4.hpp:66,
                 from W20240417-142609-5.cpp:6:
/opt/HPCCSystems/componentfiles/cl/include/eclrtl.hpp:89:54: note:   initializing argument 2 of ‘virtual void ICompiledStrRegExpr::replace(size32_t&, char*&, size32_t, const char*, size32_t, const char*) const’
   89 |     virtual void replace(size32_t & outlen, char * & out, size32_t slen, char const * str, size32_t rlen, char const * replace) const = 0;
      |                                             ~~~~~~~~~^~~
W20240417-142609-5_1.cpp: In member function ‘virtual size32_t cAc3::transform(ARowBuilder&, const void*)’:
W20240417-142609-5_1.cpp:78:49: error: cannot convert ‘UChar*’ {aka ‘short unsigned int*’} to ‘char*&’
   78 |                 regexG1->replace(vI1,vH1.refustr(),*((unsigned *)(left + 30U)),(char *)(left + 34U),1U,((char *)"\303\253"));
      |                                      ~~~~~~~~~~~^~
      |                                                 |
      |                                                 UChar* {aka short unsigned int*}
In file included from /opt/HPCCSystems/componentfiles/cl/include/eclinclude4.hpp:66,
                 from W20240417-142609-5_1.cpp:6:
/opt/HPCCSystems/componentfiles/cl/include/eclrtl.hpp:89:54: note:   initializing argument 2 of ‘virtual void ICompiledStrRegExpr::replace(size32_t&, char*&, size32_t, const char*, size32_t, const char*) const’
   89 |     virtual void replace(size32_t & outlen, char * & out, size32_t slen, char const * str, size32_t rlen, char const * replace) const = 0;
      |                                             ~~~~~~~~~^~~
W20240417-142609-5_1.cpp:90:49: error: cannot convert ‘UChar*’ {aka ‘short unsigned int*’} to ‘char*&’
   90 |                 regexM1->replace(vO1,vN1.refustr(),*((unsigned *)(left + 30U)),(char *)(left + 34U),1U,((char *)"\303\253"));
      |                                      ~~~~~~~~~~~^~
      |                                                 |
      |                                                 UChar* {aka short unsigned int*}
In file included from /opt/HPCCSystems/componentfiles/cl/include/eclinclude4.hpp:66,
                 from W20240417-142609-5_1.cpp:6:
/opt/HPCCSystems/componentfiles/cl/include/eclrtl.hpp:89:54: note:   initializing argument 2 of ‘virtual void ICompiledStrRegExpr::replace(size32_t&, char*&, size32_t, const char*, size32_t, const char*) const’
   89 |     virtual void replace(size32_t & outlen, char * & out, size32_t slen, char const * str, size32_t rlen, char const * replace) const = 0;
      |                                             ~~~~~~~~~^~~
```
Moreover the ECLCC  terminated with signal SIGSEGV, Segmentation fault and generated core files during to compile regexfindset.ecl.

**Comment by ghalliday** [comment]
@dcamper currently failing a regression test

**Comment by dcamper** [comment]
@ghalliday ready for a recheck when you have time.

File: ecl/hql/hqlgram2.cpp [hunk 1]
```diff
@@ -4974,6 +4977,25 @@ void HqlGram::ensureUnicode(attribute &a)
     }
 }
 
+void HqlGram::ensureUTF8(attribute &a)
+{
+    ITypeInfo *t1 = a.queryExprType();
+    if (t1 && !isUTF8Type(t1))
+    {
+        if (isStringType(t1))
```
Comment (ghalliday) [inline ecl/hql/hqlgram2.cpp:4985]: minor: is StringType() || isUnicodeType() ?

File: ecl/hql/hqlutil.cpp [hunk 2]
```diff
@@ -10212,6 +10212,15 @@ IHqlExpression * convertSetToExpression(bool isAll, size32_t len, const void * p
                 presult += numUChars;
             };
             break;
+        case type_utf8:
+            while (presult < presult_end)
+            {
+                const size32_t numUChars = *((size32_t *) presult);
+                presult += sizeof(size32_t);
+                results.append(*createConstant(createUtf8Value((unsigned)numUChars, (const char*)presult, makeUtf8Type(numUChars, NULL))));
+                presult += numUChars;
```
Comment (ghalliday) [inline ecl/hql/hqlutil.cpp:10221]: I think this should be.

presult += rtlUtf8Size(numUChars, presult)

Is there a test case where the set returns multi-byte utf-8 characters?

File: ecl/hql/hqlutil.cpp [hunk 3]
```diff
@@ -10709,7 +10718,12 @@ IException * checkRegexSyntax(IHqlExpression * expr)
         {
             try
             {
-                if (isUnicodeType(expr->queryType()))
+                if (isUTF8Type(expr->queryType()))
+                {
+                    ICompiledStrRegExpr * compiled = rtlCreateCompiledU8StrRegExpr(rtlUtf8Size(value->getSize(), value->queryValue()), (const char *)value->queryValue(), false);
```
Comment (ghalliday) [inline ecl/hql/hqlutil.cpp:10723]: getSize() is the size of the string, so no need to call rtlUtf8Size again.  I think possibly this should be a call to rtlUtf8Length().  Alternatively value->queryType()->getStringLen().

File: rtl/eclrtl/eclregex.cpp [hunk 4]
```diff
@@ -183,19 +191,37 @@ class CCompiledStrRegExpr : implements ICompiledStrRegExpr
...
+        }
+    }
+
+    CCompiledStrRegExpr(int _regexLength, const char * _regex, bool _isCaseSensitive, bool _enableUTF8)
+    : isUTF8Enabled(_enableUTF8)
+    {
+        int errNum = 0;
+        PCRE2_SIZE errOffset;
+        uint32_t options = ((_isCaseSensitive ? 0 : PCRE2_CASELESS) | (_enableUTF8 ? PCRE2_UTF : 0));
+        size32_t regexLength = (isUTF8Enabled ? rtlUtf8Size(_regexLength, _regex) : _regexLength);
```
Comment (ghalliday) [inline rtl/eclrtl/eclregex.cpp:218]: regexSize a better variable name for consistency

File: rtl/eclrtl/eclregex.cpp [hunk 5]
```diff
@@ -286,17 +319,18 @@ class CCompiledStrRegExpr : implements ICompiledStrRegExpr
...
-                * (size32_t *) outData = matchLen;
-                memcpy(outData + sizeof(size32_t), matchStart, matchLen);
-                outBytes += matchLen + sizeof(size32_t);
+                * (size32_t *) outData = (isUTF8Enabled ? rtlUtf8Length(matchSize, matchStart) : matchSize);
+                memcpy(outData + sizeof(size32_t), matchStart, matchSize);
+                outBytes += matchSize + sizeof(size32_t);
 
                 // Update offset
-                offset += matchLen + 1;
+                offset += matchSize + 1;
```
Comment (ghalliday) [inline rtl/eclrtl/eclregex.cpp:333]: Not new code, but is this +1 correct?   E.g. it suggests that matching "ab" against "abab" would only get the first match.

**Review by ghalliday — COMMENTED** [review]
@dcamper generally looks good.  Some of this code is >15 years old.  I wouldn't code it in the same way today - but copying the style of the existing code makes sense, so I haven't commented in general when it could have been cleaned up.

A few fairly minor comments, but looks like it is close to ready.

File: ecl/hql/hqlgram2.cpp (hunk 1 above)
Comment (dcamper) [reply ecl/hql/hqlgram2.cpp:4985]: Good catch; changed.

isUnicodeType() tests for UTF-8 and there seems to be a lot of code that leverages the implied "UTF-8 is a subset of Unicode" logic.  This is debatable, but not something to fix here.

File: ecl/hql/hqlutil.cpp (hunk 3 above)
Comment (dcamper) [reply ecl/hql/hqlutil.cpp:10723]: Fixed.

File: rtl/eclrtl/eclregex.cpp (hunk 4 above)
Comment (dcamper) [reply rtl/eclrtl/eclregex.cpp:218]: Agreed; changed.

File: rtl/eclrtl/eclregex.cpp (hunk 5 above)
Comment (dcamper) [reply rtl/eclrtl/eclregex.cpp:333]: Better yet:

offset = ovector[1];

File: ecl/hql/hqlutil.cpp (hunk 2 above)
Comment (dcamper) [reply ecl/hql/hqlutil.cpp:10221]: Also, added test cases for multibyte Unicode and UTF-8 REGEXFINDSET.  This uncovered some issues in both folded and unfolded code, cascading some further changes that will be in the next commit.

**Review by ghalliday — APPROVED** [review]
Looks good .  Please squash and I will merge.

**Comment by dcamper** [comment]
@ghalliday squashed, please merge.  Thanks!
